robot/impl/optional_dependencies_test.go

```diff
@@ -843,3 +847,163 @@ func TestOptionalDependenciesCycles(t *testing.T) {
 		test.That(t, moc2.otherMOC, test.ShouldBeNil)
 	}
 }
+
+func TestModularOptionalDependenciesCycles(t *testing.T) {
+	// This test is a copy of TestOptionalDependenciesCycles, but it also ensures that
+	// modular resources can optionally depend upon each other _and_ lack a `Reconfigure`
+	// method (leverage `resource.AlwaysRebuild`).
+
+	logger, logs := logging.NewObservedTestLogger(t)
+	ctx := context.Background()
+
+	lr := setupLocalRobot(t, ctx, &config.Config{}, logger)
+
+	optionalDepsModulePath := testutils.BuildTempModule(t, "examples/customresources/demos/optionaldepsmodule")
+
+	mutualOptionalChildModel := resource.NewModel("acme", "demo", "moc")
+	mocName := generic.Named("moc")
+	mocName2 := generic.Named("moc2")
+
+	// Reconfigure the robot to have a mutual optional child component that is missing its
+	// mutual.
+	cfg := config.Config{
+		Modules: []config.Module{
+			{
+				Name:    "optional-deps",
+				ExePath: optionalDepsModulePath,
+			},
+		},
+		Components: []resource.Config{
+			{
+				Name:  mocName.Name,
+				API:   generic.API,
+				Model: mutualOptionalChildModel,
+				Attributes: rutils.AttributeMap{
+					"other_moc": mocName2.Name,
+				},
+			},
+		},
+	}
+	// Ensure here and for all configs below before `Reconfigure`ing to make sure optional
+	// dependencies are calculated (`ImplicitOptionalDependsOn` is filled in).
+	test.That(t, cfg.Ensure(false, logger), test.ShouldBeNil)
+	lr.Reconfigure(ctx, &cfg)
+
+	{ // Assertions
+		// Assert that the mutual optional child component built successfully (optional
+		// dependency on non-existent 'moc2' did not cause a failure to build).
+		mocRes, err := lr.ResourceByName(mocName)
+		test.That(t, err, test.ShouldBeNil)
+
+		// Assert that the mutual optional logged its inability to get 'moc2' from
+		// dependencies _twice_. The first is from construction of the resource, and the
+		// second is from reconstruction of the resource (always rebuild) due to a call to
+		// `updateWeakAndOptionalDependents` directly after `completeConfig`.
+		msgNum := logs.FilterMessageSnippet("could not get other MOC").Len()
+		test.That(t, msgNum, test.ShouldEqual, 2)
+
+		// Assert that, on the component itself, `otherMOC` remains unset.
+		doCommandResp, err := mocRes.DoCommand(ctx, map[string]any{"command": "other_moc_state"})
+		test.That(t, err, test.ShouldBeNil)
+		test.That(t, doCommandResp, test.ShouldResemble, map[string]any{"other_moc_state": "unset"})
+	}
+
+	// Reconfigure the robot to have the other MOC.
+	cfg = config.Config{
+		Modules: []config.Module{
+			{
+				Name:    "optional-deps",
+				ExePath: optionalDepsModulePath,
+			},
+		},
+		Components: []resource.Config{
+			{
+				Name:  mocName.Name,
+				API:   generic.API,
+				Model: mutualOptionalChildModel,
+				Attributes: rutils.AttributeMap{
+					"other_moc": mocName2.Name,
+				},
+			},
+			{
+				Name:  mocName2.Name,
+				API:   generic.API,
```

Comment:
Given we're on a path to removing API -- this isn't a critique, just a thought experiment for further understanding.

So we're seeing that creating `mocName2` does not reconfigure `mocName` (but the client handle on `mocName` presumably will start working under the hood).

If `mocName2` were to change its API to say the motor API, would that cause a reconstruction/reconfigure of `mocName`? Or would `mocName` be left with a "stale"[1] client object?

[1] The underlying connection and resource name never changes. But the set of methods that could be called "does not update" when it theory it ought to.

Reply:
in the current state, I do think that `mocName` will get reconfigured. `mocName2` will get closed and then re-added, which should prompt `mocName` to get reconfigured as well. but not super high confidence

Reply:
> So we're seeing that creating mocName2 does not reconfigure mocName (but the client handle on mocName presumably will start working under the hood).

For the new test I've added, when we add `moc2`, `moc` _will technically get rebuilt_ module-side (it has no `reconfigure` method but is able to reconstruct in response to a `ReconfigureResourceRequest`) and given a gRPC client to `moc2`. In its initial construction `moc` was given _no_ gRPC client (`deps` was empty when passed to constructor). Similary, when `moc` is removed from the config, `moc2` _will get rebuilt_ to pass in _no_ gRPC client (`deps` will be empty). So, I don't think any gRPC client "starts" or "stops" working under the hood with this test. Maybe I'm misunderstanding your statement here.